cura/Machines/MaterialManager.py

```diff
@@ -0,0 +1,324 @@
+from collections import defaultdict, OrderedDict
+from typing import Optional
+
+from PyQt5.Qt import QTimer, QObject, pyqtSignal
+
+from UM.Logger import Logger
+from UM.Settings import ContainerRegistry
+
+from cura.Machines.ContainerNode import ContainerNode
+
+
+class MaterialGroup:
+    __slots__ = ("name", "root_material_node", "derived_material_node_list")
+
+    def __init__(self, name: str):
+        self.name = name
+        self.root_material_node = None
+        self.derived_material_node_list = []
+
+    def __str__(self) -> str:
+        return "%s[%s]" % (self.__class__.__name__, self.name)
+
+
+class MaterialNode(ContainerNode):
```

Comment:
This violates our code style: One class per file and the class should have the same name as the file.

Reply:
Moved to a separate file in c5d443109